### DISCUSSION TITLE
App onStart event

### DISCUSSION THREAD

**Discussion by nemanjam** [withastro/roadmap#1005] [Proposal]
# Summary

`onStart` event that runs only once in both development and production modes that can be used for example for logging the initial configuration or other jobs that should run once on app start.

# Background & Motivation

In my example I am parsing all environment variables and static configuration, validating them with Zod and printing them out on app start to be clear what config app used in different environments and nothing is undefined or incorrect.

```ts
import { NODE_ENV as NODE_ENV_STRING, PREVIEW_MODE } from 'astro:env/server';
import { SITE_URL } from 'astro:env/client';

import { configServerSchema } from '@/schemas/config';
import { configClientSchema } from '@/schemas/config';

import { prettyPrintObject } from '@/utils/log';
import { validateData } from '@/utils/validation';

import type { ConfigServerType } from '@/types/config';
import type { ConfigClientType } from '@/types/config';

// validate server config

// cast type
const NODE_ENV = NODE_ENV_STRING as ConfigServerType['NODE_ENV'];

/** SSG - all env vars are build time only. */
const configServerData: ConfigServerType = { NODE_ENV, PREVIEW_MODE };

export const CONFIG_SERVER = validateData(configServerData, configServerSchema);


// validate client config

const configClientData: ConfigClientType = {
  /** all urls without '/' */
  SITE_URL,
  SITE_TITLE: 'My blog',
  // other client static config
};

export const CONFIG_CLIENT = validateData(configClientData, configClientSchema);

// print merged config
export const MERGED_CONFIG = { ...CONFIG_SERVER, ...CONFIG_CLIENT };

// run this in onStart handler
prettyPrintObject(MERGED_CONFIG, 'parsed CONFIG');

```

# Goals

Currently I can just call print function in global scope, but I have no clear idea when it will run and how many time and if config is expected to be loaded properly at that point. In practice I can see it mostly runs twice in development. Or I can try run it in some base layout but then it will run on every page load which I don't want. For example Express has events for start, error, etc. and for Astro even for static builds can be useful to log configs when building the app or even at runtime.

# Example

Not only `onStart` but additional events can be emitted like `onError`, `onExit`, `onConfigLoaded`, etc.

**Comment by aretrace** [comment]
yes

**Comment by edvinasjurele** [comment]
Hey, we do have a use case where we need to run some initialization stuff, like Redis connections when server starts. We had been using the `middlware/index.ts` file before because it was running during initialization and this worked on 4.x, but was removed in v5.0.5 with https://github.com/withastro/astro/pull/12707/files

Given that, since the middleware runs with the first GET request (either page or API endpoint) and not on server init, we get Redis initialized later too.

**Comment by ematipico** [comment]
There's no concept of "app start" inside an Astro application.

Or, to be more precise, it's not possible to translate a concept like this in serverless environments.

It seems you want to have such a thing just for Node.js, which could be solved using the middleware mode of the Node.js adapter

**Reply by edvinasjurele** [reply]
You are right, this is rather about node, and you are right, we run it as middleware:
```
output: 'server',
  adapter: node({
    mode: 'middleware',
  }),
  ```
  
 So we have custom Fastify server https://docs.astro.build/en/guides/integrations-guide/node/#middleware. I am aware that we could do extra stuff inside the server, but we would like to have all app logic inside `./dist/server/entry.mjs` which is the outcome of Astro build, so probably this is rather about `@astrojs/node` 🙏 

I see there is another discussion which suggests exactly that https://github.com/withastro/roadmap/discussions/967

**Comment by edvinasjurele** [comment]
Duplicate discussion with https://github.com/withastro/roadmap/discussions/967

**Comment by smnbbrv** [comment]
Posting as duplicate to https://github.com/withastro/roadmap/discussions/967

Hi guys, as a one who was suffering from this problem, I invested some time into a solution.

https://github.com/smnbbrv/astroscope/tree/main/packages/boot

Please let me know what you think.
